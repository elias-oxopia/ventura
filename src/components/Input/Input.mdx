---
name: Input
menu: Components
route: /input
---

import { useState } from 'react';
import { Playground, Props } from 'docz';
import Input from './Input.tsx';
import styles from './Input.module.css';

# Input

A Input can be used to let the user insert text using key strokes. A crucial form item for building interactive apps.

## Best practices

- Inputs should be wrapped in a FormField component or should be preceded by another component component serving as a label // TODO Link to FormField

## Properties

<Props of={Input} />

## Examples

### Basic usage

<Playground>
  <Input name="firstname" />
</Playground>

### Listening for changes

<Playground>
  <Input name="logger" onChange={(event) => console.log(`LOGGER: ${event.target.value}`)} />
</Playground>

### Email address

<Playground>
  <Input name="email" placeholder="Email address" type="email" />
</Playground>

### Password

<Playground>
  <Input name="password" placeholder="Password" type="password" />
</Playground>

### Limited amount of characters

<Playground>
  {/* .countryCodeInput { text-align: center; width: 30px; } */}
  <Input name="countryCode" maxLength={2} className={styles.countryCodeInput} placeholder="XX" />
</Playground>

### Error message

<Playground>
  <Input name="username" errorMessage="The username contains invalid characters" />
</Playground>

### Format a serial code

<Playground>
  {() => {
    const [serialCode, setSerialCode] = useState('AAAA-BBBB-CCCC-DDDD');
    const formattedSerialCode = (
      serialCode
        .split('-')
        .join('')
        .match(/.{1,4}/g) || []
    )
      .join('-')
      .toUpperCase();
    return (
      <Input
        name="serialCode"
        value={formattedSerialCode}
        onChange={(event) => setSerialCode(event.target.value)}
        maxLength={19}
      />
    );
  }}
</Playground>
