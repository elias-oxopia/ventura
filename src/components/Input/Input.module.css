.container {
  width: 100%;
}

.input {
  font: inherit;
  box-sizing: border-box;
  width: inherit;
  background-color: var(--color-neutral-1);
  border-radius: var(--border-radius-small);
  border: 1px solid var(--color-neutral-4);
  padding: 6.5px 14px;
}

.input::placeholder {
  color: var(--color-neutral-6);
}

.input:focus {
  outline: none;
  box-shadow: 0 0 0 3px var(--color-secondary-focus);
}

.containsError {
  border-color: var(--color-error);
}

.containsError:focus {
  box-shadow: 0 0 0 3px var(--color-error-light);
}

.errorMessage {
  display: block;
  margin-top: 4px;
  color: var(--color-error);
}
