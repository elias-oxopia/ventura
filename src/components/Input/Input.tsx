import React, { ChangeEventHandler, FocusEventHandler } from 'react';
import classNames from 'classnames';

import cssReset from '../../css-reset.module.css';
import styles from './Input.module.css';

interface Props {
  name: string;
  value?: string;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  onBlur?: FocusEventHandler<HTMLInputElement>;
  placeholder?: string;
  type?: 'text' | 'email' | 'password';
  errorMessage?: string;
  spellCheck?: boolean;
  autoComplete?: boolean;
  maxLength?: number;
  className?: string;
  isInvalid?: boolean;
}

const Input = React.forwardRef<HTMLInputElement, Props>(
  (
    {
      name,
      value,
      onChange,
      onBlur,
      placeholder,
      type = 'text',
      isInvalid = false,
      spellCheck = false,
      autoComplete = false,
      maxLength,
      className,
    }: Props,
    ref,
  ) => {
    const inputClassNames = classNames(cssReset.ventura, styles.input, {
      [styles.containsError]: Boolean(isInvalid),
      className,
    });

    return (
      <input
        name={name}
        value={value}
        onChange={onChange}
        placeholder={placeholder}
        type={type}
        spellCheck={spellCheck}
        autoComplete={autoComplete ? 'on' : 'off'}
        maxLength={maxLength}
        ref={ref}
        className={inputClassNames}
        data-testid={`input-${name}`}
        onBlur={onBlur}
      />
    );
  },
);

export default Input;
