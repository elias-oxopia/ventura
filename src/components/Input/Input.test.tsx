import React, { ComponentProps } from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Input from './Input';

describe('Input', () => {
  const defaultButtonProps: ComponentProps<typeof Input> = {
    name: 'firstname',
  };

  test('default snapshot', () => {
    const component = <Input {...defaultButtonProps} />;
    const { asFragment } = render(component);
    expect(asFragment()).toMatchSnapshot();
  });

  test('value prop', () => {
    const component = <Input {...defaultButtonProps} value="Robin" onChange={() => {}} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('value')).toBe('Robin');
  });

  test('onChange prop', async () => {
    const onChangeFn = jest.fn();
    const firstName = 'Robin';
    const component = <Input {...defaultButtonProps} onChange={onChangeFn} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    await userEvent.type(inputElement, firstName);

    expect(onChangeFn).toHaveBeenCalledTimes(firstName.length);
    expect((inputElement as any).value).toBe(firstName);
  });

  test('placeholder prop', () => {
    const component = <Input {...defaultButtonProps} placeholder="First name" />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('placeholder')).toBe('First name');
  });

  test('type prop', () => {
    const component = <Input {...defaultButtonProps} type="password" />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('type')).toBe('password');
  });

  test('errorMessage prop', () => {
    const component = <Input {...defaultButtonProps} isInvalid />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.classList).toContain('containsError');
  });

  test('spellCheck prop', () => {
    const component = <Input {...defaultButtonProps} spellCheck />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.hasAttribute('spellcheck')).toBe(true);
    expect(inputElement.getAttribute('spellcheck')).not.toBe(false);
  });

  test('autoComplete prop', () => {
    const component = <Input {...defaultButtonProps} autoComplete />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('autocomplete')).toBe('on');
  });

  test('maxLength prop', () => {
    const maxLength = 3;
    const component = <Input {...defaultButtonProps} maxLength={3} />;
    const { getByTestId } = render(component);
    const inputElement = getByTestId('input-firstname');

    expect(inputElement.getAttribute('maxlength')).toBe(maxLength.toString());
  });

  test('className prop', () => {
    const className = 'center';
    const component = <Input {...defaultButtonProps} className={className} />;
    const { getByTestId } = render(component);
    const containerElement = getByTestId('container');

    const renderedClassNames = containerElement.className.split(' ');
    expect(renderedClassNames).toContain(className);
    // className in prop should be the last in the row
    expect(renderedClassNames.indexOf(className)).toBe(renderedClassNames.length - 1);
  });
});
