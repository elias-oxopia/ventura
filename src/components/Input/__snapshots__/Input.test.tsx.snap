// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input default snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="container"
  >
    <input
      autocomplete="off"
      class="input"
      data-testid="input-firstname"
      name="firstname"
      spellcheck="false"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
